### PR TITLE
feat: Добавлена новая диагностика `ServerCallsInFormEventsDiagnostic`

### LINKED ISSUE #3473: [NEW]
Добавить диагностику, которая проверяет, что в событиях `ПриАктивизацииСтроки` и `НачалоВыбора` нет методов исполняемых на сервере.
Если метод находится внутри модуля формы, то необходимо проверить директиву метода, если на сервере, то это ошибка.

## Ссылка на источник, подтверждающее нарушение либо обоснование наличия проблемы
Image

### Тип [Статья на русском](https://1c-syntax.github.io/bsl-language-server/contributing/DiagnosticTypeAndSeverity)

* [ ] :ant: Ошибка
* [ ] :cop: Уязвимость
* [ ] :guardsman: Потенциальная уязвимость
* [ ] :poop: Качество кода
* [ ] :trollface: Другое

### Важность [Статья на русском](https://1c-syntax.github.io/bsl-language-server/contributing/DiagnosticTypeAndSeverity)

* [ ] :broken_heart: Блокирующая / Blocker
* [ ] :heart: Критическая / Critical
* [ ] :yellow_heart: Важная / Major
* [ ] :blue_heart: Незначительная / Minor
* [ ] :green_heart: Информационная / Info
* [ ] :revolving_hearts: Другое

### Тэги [Статья на русском](https://1c-syntax.github.io/bsl-language-server/contributing/DiagnosticTag)

* [ ] `STANDARD` - "Нарушение стандартов 1С"
* [ ] `LOCKINOS` - "Не будет работать в другой ОС"
* [ ] `SQL` - "Проблема с запросом"
* [ ] `PERFORMANCE` - "Проблема производительности"
* [ ] `BRAINOVERLOAD` - "Непонятный код"
* [ ] `BADPRACTICE` - "Плохая практика программирования"
* [ ] `CLUMSY` - "Излишние действия"
* [ ] `DESIGN` - "Ошибка в проектировании"
* [ ] `SUSPICIOUS` - "Подозрительный код"
* [ ] `UNPREDICTABLE` - "Непредсказуемо работающий код"
* [ ] `DEPRECATED` - "Устаревшая функциональность"
* [ ] `ERROR` - "Ошибочная конструкция"
* [ ] `LOCALIZE` - "Проблемы локализации"

## Дополнительная информация

[Серверные вызовы, которые нельзя вызывать](https://infostart.ru/1c/articles/1225834/)

### PR DESCRIPTION
**Pull request by johnnyshut**
## Описание

Добавлена новая диагностика `ServerCallsInFormEventsDiagnostic` для обнаружения серверных вызовов в событиях `ПриАктивизацииСтроки` (OnActivateRow) и `НачалоВыбора` (OnStartChoice) в модулях форм.

### Внесенные изменения:

1. **Новая диагностика** - `ServerCallsInFormEventsDiagnostic.java`
   - Обнаруживает вызовы серверных процедур в клиентских событиях форм
   - Работает только в модулях типа `FormModule`
   - Поддерживает как русские, так и английские названия событий

2. **Локализация** — добавлены файлы сообщений на русском и английском языках
   - `ServerCallsInFormEventsDiagnostic_ru.properties`
   - `ServerCallsInFormEventsDiagnostic_en.properties`

3. **Документация** — создана документация для диагностики
   - `docs/diagnostics/ServerCallsInFormEvents.md` (русский)
   - `docs/en/diagnostics/ServerCallsInFormEvents.md` (английский)
   - Включает описание проблемы, примеры и ссылку на статью Infostart

4. **Тесты** — добавлены comprehensive unit тесты
   - `ServerCallsInFormEventsDiagnosticTest.java` - unit тесты
   - `ServerCallsInFormEventsDiagnostic.bsl` - BSL тестовый файл
   - Покрывают все сценарии использования

### Технические детали:
- Диагностика использует visitor pattern для обхода AST
- Проверяет тип модуля перед выполнением
- Отслеживает состояние входа/выхода из запрещенных событий
- Срабатывает только на вызовы серверных процедур в клиентских событиях

## Связанные задачи

Closes #3473

### Общие

- [x] Ветка PR обновлена из develop
- [x] Отладочные, закомментированные и прочие, не имеющие смысла участки кода удалены
- [x] Изменения покрыты тестами
- [x] Обязательные действия перед коммитом выполнены (запускал команду `gradlew precommit`)

### Для диагностик

- [x] Описание диагностики заполнено для обоих языков (присутствуют файлы для обоих языков, для русского заполнено все подробно, перевод на английский можно опустить)

### Ссылки на источники:
- [Статья Infostart](https://infostart.ru/1c/articles/1225834/) - источник требований к диагностике

### Примеры использования:

**Неправильно (вызовет диагностику):**
```bsl
Процедура ПриАктивизацииСтроки(Элемент, ВыбраннаяСтрока, Поле, НовоеЗначение, СтандартнаяОбработка)
    // Ошибка: серверный вызов в клиентском событии
    ТаблицаФормыПриАктивизацииСтрокиНаСервере();
    СтандартнаяОбработка = Ложь;
КонецПроцедуры
```

**Правильно:**
```bsl
Процедура ПриАктивизацииСтроки(Элемент, ВыбраннаяСтрока, Поле, НовоеЗначение, СтандартнаяОбработка)
    // Правильно: только клиентская логика
    СтандартнаяОбработка = Ложь;
КонецПроцедуры
```

### Тестирование:
- Все unit тесты проходят успешно
- Диагностика корректно работает только в модулях форм
- Поддерживает как русские, так и английские названия событий

---

**Примечание:** Перед отправкой PR необходимо выполнить команду `gradlew precommit` для проверки всех обязательных действий.

## Summary by CodeRabbit

* **New Features**
  * Diagnostic that flags server procedure calls inside client form events (OnActivateRow / OnStartChoice and Russian equivalents).

* **Documentation**
  * Added RU/EN docs describing the diagnostic with incorrect vs correct examples and references.

* **Tests**
  * Added comprehensive unit tests covering Russian/English names, case-insensitivity, edge cases, multiple calls and directive handling.

* **Configuration**
  * New schema/parameter to enable/adjust the diagnostic.

* **Localization**
  * Added English and Russian diagnostic messages.

### CODE REVIEW

File: src/main/java/com/github/_1c_syntax/bsl/languageserver/diagnostics/ServerCallsInFormEventsDiagnostic.java [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+              (compilerDirective.get() == com.github._1c_syntax.bsl.languageserver.context.symbol.annotations.CompilerDirectiveKind.AT_SERVER ||
+               compilerDirective.get() == com.github._1c_syntax.bsl.languageserver.context.symbol.annotations.CompilerDirectiveKind.AT_SERVER_NO_CONTEXT)) {
+            diagnosticStorage.addDiagnostic(ctx,
+              info.getMessage("message", methodName));
+          }
+        }
+      }
+    }
+    return super.visitGlobalMethodCall(ctx);
+  }
```
[the remark this answers is not in the document]
Comment (johnnyshut) [reply src/main/java/com/github/_1c_syntax/bsl/languageserver/diagnostics/ServerCallsInFormEventsDiagnostic.java:162]: Напиши проще и по-русски, пожалуйста

File: src/main/java/com/github/_1c_syntax/bsl/languageserver/diagnostics/ServerCallsInFormEventsDiagnostic.java [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+    super();
+  }
+
+  /**
+   * Обрабатывает посещение подпрограммы (процедуры или функции).
+   * 
+   * Проверяет, является ли подпрограмма событием ПриАктивизацииСтроки или НачалоВыбора,
+   * и устанавливает флаг для отслеживания контекста.
+   * 
+   * @param ctx контекст подпрограммы
```
Comment (theshadowco) [inline src/main/java/com/github/_1c_syntax/bsl/languageserver/diagnostics/ServerCallsInFormEventsDiagnostic.java]: эммм? подпрограммы?

File: src/main/java/com/github/_1c_syntax/bsl/languageserver/diagnostics/ServerCallsInFormEventsDiagnostic.java [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+   * 
+   * @param ctx контекст вызова глобального метода
+   * @return результат обхода дерева
+   */
+  @Override
+  public ParseTree visitGlobalMethodCall(BSLParser.GlobalMethodCallContext ctx) {
+    // Проверяем, что мы в модуле формы
+    if (documentContext.getModuleType() != ModuleType.FormModule) {
+      return super.visitGlobalMethodCall(ctx);
+    }
```
Comment (theshadowco) [inline src/main/java/com/github/_1c_syntax/bsl/languageserver/diagnostics/ServerCallsInFormEventsDiagnostic.java]: зачем? по настройке аннотации иного быть и не может

File: src/main/java/com/github/_1c_syntax/bsl/languageserver/diagnostics/ServerCallsInFormEventsDiagnostic.java [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+    
+    if (isInForbiddenEvent) {
+      // Проверяем, вызывается ли серверная процедура из клиентского события
+      String methodName = ctx.methodName().getText();
```
Comment (theshadowco) [inline src/main/java/com/github/_1c_syntax/bsl/languageserver/diagnostics/ServerCallsInFormEventsDiagnostic.java]: var вместо явного указания String

File: src/main/java/com/github/_1c_syntax/bsl/languageserver/diagnostics/ServerCallsInFormEventsDiagnostic.java [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+      return false;
+    }
+    
+    var methodSymbol = methodSymbolOpt.get();
+    var directiveOpt = methodSymbol.getCompilerDirectiveKind();
+    var serverDirectives = EnumSet.of(
+      CompilerDirectiveKind.AT_SERVER,
+      CompilerDirectiveKind.AT_SERVER_NO_CONTEXT,
+      CompilerDirectiveKind.AT_CLIENT_AT_SERVER,
+      CompilerDirectiveKind.AT_CLIENT_AT_SERVER_NO_CONTEXT
```
Comment (theshadowco) [inline src/main/java/com/github/_1c_syntax/bsl/languageserver/diagnostics/ServerCallsInFormEventsDiagnostic.java]: стоит эту коллекцию хранить статично

File: src/main/java/com/github/_1c_syntax/bsl/languageserver/diagnostics/ServerCallsInFormEventsDiagnostic.java [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+    if (methodSymbolOpt.isEmpty()) {
+      return false;
+    }
+    
+    var methodSymbol = methodSymbolOpt.get();
+    var directiveOpt = methodSymbol.getCompilerDirectiveKind();
+    var serverDirectives = EnumSet.of(
+      CompilerDirectiveKind.AT_SERVER,
+      CompilerDirectiveKind.AT_SERVER_NO_CONTEXT,
+      CompilerDirectiveKind.AT_CLIENT_AT_SERVER,
```
Comment (theshadowco) [inline src/main/java/com/github/_1c_syntax/bsl/languageserver/diagnostics/ServerCallsInFormEventsDiagnostic.java]: а чем провинился метод, доступный на клиенте и сервере? серверного вызова не будет по определению

File: docs/diagnostics/ServerCallsInFormEvents.md [hunk 7]
```diff
@@ -0,0 +1,43 @@
+# Серверные вызовы в событиях форм (ServerCallsInFormEvents)
+
+<!-- Блоки выше заполняются автоматически, не трогать -->
+## Описание диагностики
+<!-- Описание диагностики заполняется вручную. Необходимо понятным языком описать смысл и схему работу -->
+
+В событиях `ПриАктивизацииСтроки` и `НачалоВыбора` не должно быть вызовов серверных процедур. Эти события должны выполняться только на клиенте.
```
Comment (theshadowco) [inline docs/diagnostics/ServerCallsInFormEvents.md:7]: стоит добавить, что диагностируются только серверные вызовы методов ЭТОЙ формы, т.е. вызов `МойМодульСервер.МойСерверныйМетод` не будет ошибкой

File: src/main/resources/com/github/_1c_syntax/bsl/languageserver/diagnostics/ServerCallsInFormEventsDiagnostic_en.properties [hunk 8]
```diff
@@ -0,0 +1,11 @@
+# ServerCallsInFormEventsDiagnostic diagnostic messages in English
+
+# Diagnostic name
```
Comment (theshadowco) [inline src/main/resources/com/github/_1c_syntax/bsl/languageserver/diagnostics/ServerCallsInFormEventsDiagnostic_en.properties]: эти комментарии здесь не нужны, стоит оформить по образу и подобию соседних

File: src/main/resources/com/github/_1c_syntax/bsl/languageserver/diagnostics/ServerCallsInFormEventsDiagnostic_ru.properties [hunk 9]
```diff
@@ -0,0 +1,11 @@
+# Сообщения диагностики ServerCallsInFormEventsDiagnostic на русском языке
+
+# Название диагностики
```
Comment (theshadowco) [inline src/main/resources/com/github/_1c_syntax/bsl/languageserver/diagnostics/ServerCallsInFormEventsDiagnostic_ru.properties]: см выше

File: src/test/java/com/github/_1c_syntax/bsl/languageserver/diagnostics/ServerCallsInFormEventsDiagnosticTest.java [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+class ServerCallsInFormEventsDiagnosticTest extends AbstractDiagnosticTest<ServerCallsInFormEventsDiagnostic> {
+
+  ServerCallsInFormEventsDiagnosticTest() {
+    super(ServerCallsInFormEventsDiagnostic.class);
+  }
+
+  @Test
+  void test() {
+    // Создаем контекст для модуля формы с тестовым кодом
+    String fileContent = """
```
Comment (theshadowco) [inline src/test/java/com/github/_1c_syntax/bsl/languageserver/diagnostics/ServerCallsInFormEventsDiagnosticTest.java]: Не используем такой формат фикстур: фикстуры идут как ресурсные файлы. 
Кроме того, файл модуля есть, зачем эти строки - не ясно

**Comment by johnnyshut** [comment]
@theshadowco исправил замечания

File: src/main/java/com/github/_1c_syntax/bsl/languageserver/diagnostics/ServerCallsInFormEventsDiagnostic.java [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+   * Обрабатывает вызов процедуры или функции.
+   * 
+   * Проверяет, является ли процедура или функция событием ПриАктивизацииСтроки или НачалоВыбора,
+   * и устанавливает флаг для отслеживания контекста.
+   * 
+   * @param ctx контекст процедуры или функции
+   * @return результат обхода дерева
+   */
+  @Override
+  public ParseTree visitSub(BSLParser.SubContext ctx) {
```
Comment (nixel2007) [inline src/main/java/com/github/_1c_syntax/bsl/languageserver/diagnostics/ServerCallsInFormEventsDiagnostic.java]: 1. Здесь больше подходит AbstractListenerDiagnostic, так как тебе надо сбрасывать флаги на выходе.
2. Для неподходящих процедур нет смысла дёргать super, это будет в холостую гонять визитор по дереву. Для таких случаев надо сразу возвращать ctx.
3. Вместо разбора дерева выражений (GlobalMethodCall) можно через ReferenceIndex посмотреть все исходящие ссылки из текущего метода. Тогда по дереву вглубь ходить не надо вообще. 
4. Возможно проще и быстрее будет сделать AbstractSymbolSymbolTreeDiagnostic, отвязавшись от ParseTree полностью
5. Или пойти ещё дальше, и сделать AbstractDiagnostic, получить все method symbols из symbol tree, отфильтровать нужные и их проверить. Но это уже на твоё усмотрение.

Comment (johnnyshut) [reply src/main/java/com/github/_1c_syntax/bsl/languageserver/diagnostics/ServerCallsInFormEventsDiagnostic.java]: Посмотри теперь на реализацию. И запусти, пожалуйста, еще раз сборку, один flow упал, но вроде ошибка просто про то, что файл не удалось прочитать.

Comment (nixel2007) [reply src/main/java/com/github/_1c_syntax/bsl/languageserver/diagnostics/ServerCallsInFormEventsDiagnostic.java]: Сборка теперь сама будет запускаться от твоих pr, ты теперь контрибьютор проекта.

Comment (johnnyshut) [reply src/main/java/com/github/_1c_syntax/bsl/languageserver/diagnostics/ServerCallsInFormEventsDiagnostic.java]: Готово)))

File: src/main/resources/com/github/_1c_syntax/bsl/languageserver/diagnostics/ServerCallsInFormEventsDiagnostic_ru.properties [hunk 12]
```diff
@@ -0,0 +1,3 @@
+diagnosticName=Серверные вызовы в событиях форм
+diagnosticMessage=В событиях ПриАктивизацииСтроки и НачалоВыбора не должно быть вызовов серверных процедур. Процедура "{0}" выполняется на сервере, что может привести к проблемам.
+message=В событиях ПриАктивизацииСтроки и НачалоВыбора не должно быть вызовов серверных процедур. Процедура "{0}" выполняется на сервере, что может привести к проблемам.
```
Comment (theshadowco) [inline src/main/resources/com/github/_1c_syntax/bsl/languageserver/diagnostics/ServerCallsInFormEventsDiagnostic_ru.properties]: а зачем message? diagnosticMessage достаточно, или я что-то не вижу?

Comment (johnnyshut) [reply src/main/resources/com/github/_1c_syntax/bsl/languageserver/diagnostics/ServerCallsInFormEventsDiagnostic_ru.properties]: Использовал кастомный ключ (info.getMessage("message", methodName)), исправил на дефолтный

**Review by theshadowco — APPROVED** [review]
